core/src/main/java/com/datastax/oss/driver/internal/core/protocol/CompressorSubstitutions.java

```diff
@@ -0,0 +1,108 @@
+import com.datastax.oss.driver.api.core.context.DriverContext;
+import com.datastax.oss.protocol.internal.Compressor;
+import com.oracle.svm.core.annotate.Delete;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+import io.netty.buffer.ByteBuf;
+import java.util.function.BooleanSupplier;
+
+/**
+ * Handles GraalVM substitutions for compressors: LZ4 is only supported if we can find the native
+ * library in the classpath, and Snappy is never supported.
+ *
+ * <p>When a compressor is not supported, we delete its class, and modify {@link BuiltInCompressors}
+ * to throw an error if the user attempts to configure it.
+ */
+public class CompressorSubstitutions {
+
+  @TargetClass(value = BuiltInCompressors.class, onlyWith = Lz4Present.class)
+  public static final class BuiltInCompressorsLz4Only {
+    @Substitute
+    public static Compressor<ByteBuf> get(String name, DriverContext context) {
+      switch (name.toLowerCase()) {
+        case "lz4":
+          return new Lz4Compressor(context);
+        case "snappy":
+          throw new UnsupportedOperationException(
+              "Snappy compression is not supported for native images");
+        case "none":
+          return Compressor.none();
+        default:
+          throw new IllegalArgumentException(
+              String.format(
+                  "Unsupported compression algorithm '%s' (from configuration option %s)",
+                  name, DefaultDriverOption.PROTOCOL_COMPRESSION.getPath()));
+      }
+    }
+  }
+
+  @TargetClass(value = BuiltInCompressors.class, onlyWith = Lz4Missing.class)
+  public static final class NoBuiltInCompressors {
+    @Substitute
+    public static Compressor<ByteBuf> get(String name, DriverContext context) {
+      switch (name.toLowerCase()) {
+        case "lz4":
+          throw new UnsupportedOperationException(
+              "This native image was not built with support for LZ4 compression");
+        case "snappy":
+          throw new UnsupportedOperationException(
+              "Snappy compression is not supported for native images");
+        case "none":
+          return Compressor.none();
+        default:
+          throw new IllegalArgumentException(
+              String.format(
+                  "Unsupported compression algorithm '%s' (from configuration option %s)",
+                  name, DefaultDriverOption.PROTOCOL_COMPRESSION.getPath()));
+      }
+    }
+  }
+
+  @TargetClass(value = Lz4Compressor.class, onlyWith = Lz4Missing.class)
+  @Delete
+  public static final class DeleteLz4Compressor {}
+
+  @TargetClass(value = SnappyCompressor.class)
+  @Delete
+  public static final class DeleteSnappyCompressor {}
+
+  public static class Lz4Present implements BooleanSupplier {
+
+    private static final String LZ4_CLZ_NAME = "net.jpountz.lz4.LZ4Compressor";
+
+    @Override
+    public boolean getAsBoolean() {
+      try {
+        Class.forName(LZ4_CLZ_NAME);
```

Comment:
You could use `DependencyCheck.LZ4.isPresent()` for that, but maybe it's not a good idea to introduce dependencies in a GraalVM substitution code.

Reply:
I'm not sure it's really that big a problem, though.  If I have this right you wind up with:

```java
  public static class Lz4Present implements BooleanSupplier {

    @Override
    public boolean getAsBoolean() {
      return DependencyCheck.LZ4.isPresent();
    }
  }
```

This seems like a pretty minimal dependency... and probably worth the risk to avoid re-implementing the same thing.

Reply:
Yes, I'd rather keep the substitution code as simple as possible 👍